packages/core/src/browser/tree/tree-widget.tsx

```diff
@@ -776,6 +789,54 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
         return iconClass.concat(additionalClasses).join(' ');
     }
 
+    /**
+     * Render indent for the file tree based on the depth
+     * @param node the tree node.
+     * @param depth the depth of the tree node.
+     */
+    protected renderIndent(node: TreeNode, props: NodeProps): React.ReactNode {
+        const renderIndentGuides = this.corePreferences['workbench.tree.renderIndentGuides'];
+
+        const indentDivs: React.ReactNode[] = [];
+        let current: TreeNode | undefined = node;
+        let depth = props.depth;
+        while (current && depth) {
+            const classNames: string[] = [TREE_NODE_INDENT_GUIDE_CLASS];
+            if (renderIndentGuides !== 'none') {
+                if (this.needsActiveIndentGuideline(current)) {
+                    classNames.push('active');
+                } else {
+                    classNames.push(renderIndentGuides === 'onHover' ? 'hover' : 'always');
+                }
+            }
+            const paddingLeft = this.props.leftPadding * depth;
+            indentDivs.unshift(<div key={depth} className={classNames.join(' ')} style={{
+                paddingLeft: `${paddingLeft}px`
+            }} />);
+            current = current.parent;
+            depth--;
+        }
+        return indentDivs;
+    }
+
+    protected needsActiveIndentGuideline(node: TreeNode): boolean {
+        const parent = node.parent;
+        if (!parent || !this.isExpandable(parent)) {
+            return false;
+        }
+        if (SelectableTreeNode.isSelected(parent)) {
+            return true;
+        }
+        if (parent.expanded) {
+            for (const sibling of parent.children) {
```

Comment:
TODO: instead we should rather iterate over selected nodes and check whether its parent is a parent. There are usually not many selected nodes compare to children, i.e. test with node modules

Reply:
It works fast enough even for node modules folder.